src/ShallowWrapper.lua

```diff
@@ -0,0 +1,228 @@
+local Children = require(script.Parent.PropMarkers.Children)
+local ElementKind = require(script.Parent.ElementKind)
+local ElementUtils = require(script.Parent.ElementUtils)
+local snapshot = require(script.Parent.snapshot)
+
+local ShallowWrapper = {}
+local ShallowWrapperMetatable = {
+	__index = ShallowWrapper,
+}
+
+local function getTypeFromVirtualNode(virtualNode)
+	local element = virtualNode.currentElement
+	local kind = ElementKind.of(element)
+
+	if kind == ElementKind.Host then
+		return {
+			kind = ElementKind.Host,
+			className = element.component,
+		}
+	elseif kind == ElementKind.Function then
+		return {
+			kind = ElementKind.Function,
+			functionComponent = element.component,
+		}
+	elseif kind == ElementKind.Stateful then
+		return {
+			kind = ElementKind.Stateful,
+			component = element.component,
+		}
+	else
+		error(('shallow wrapper does not support element of kind %q'):format(tostring(kind)))
+	end
+end
+
+local function findNextVirtualNode(virtualNode, maxDepth)
+	local currentDepth = 0
+	local currentNode = virtualNode
+	local nextNode = currentNode.children[ElementUtils.UseParentKey]
+
+	while currentDepth < maxDepth and nextNode ~= nil do
+		currentNode = nextNode
+		nextNode = currentNode.children[ElementUtils.UseParentKey]
+		currentDepth = currentDepth + 1
+	end
+
+	return currentNode
+end
+
+local ContraintFunctions = {
+	kind = function(virtualNode, expectKind)
+		return ElementKind.of(virtualNode.currentElement) == expectKind
+	end,
+	className = function(virtualNode, className)
+		local element = virtualNode.currentElement
+		local isHost = ElementKind.of(element) == ElementKind.Host
+		return isHost and element.component == className
+	end,
+	component = function(virtualNode, expectComponentValue)
+		return virtualNode.currentElement.component == expectComponentValue
+	end,
+	props = function(virtualNode, propSubSet)
+		local elementProps = virtualNode.currentElement.props
+
+		for propKey, propValue in pairs(propSubSet) do
+			if elementProps[propKey] ~= propValue then
+				return false
+			end
+		end
+
+		return true
+	end,
+	hostKey = function(virtualNode, expectHostKey)
+		return virtualNode.hostKey == expectHostKey
+	end,
+}
+
+local function countChildrenOfElement(element)
+	if ElementKind.of(element) == ElementKind.Fragment then
+		local count = 0
+
+		for _, subElement in pairs(element.elements) do
+			count = count + countChildrenOfElement(subElement)
+		end
+
+		return count
+	else
+		return 1
+	end
+end
+
+local function getChildren(virtualNode, results, maxDepth)
+	if ElementKind.of(virtualNode.currentElement) == ElementKind.Fragment then
+		for _, subVirtualNode in pairs(virtualNode.children) do
+			getChildren(subVirtualNode, results, maxDepth)
+		end
+	else
+		local childWrapper = ShallowWrapper.new(
+			virtualNode,
+			maxDepth
+		)
+
+		table.insert(results, childWrapper)
+	end
+end
+
+local function filterProps(props)
```

Comment:
it could be simpler to just not filter the props in the shallow at all

Reply:
I want remove the Roact.Children key from the props when it's there. That is because it makes it easier to assert your props. Intuitively, it matches what you put in the props argument of Roact.createElement.